Assets/Mirror/Runtime/Compression.cs

```diff
@@ -0,0 +1,247 @@
+using UnityEngine;
+
+namespace Mirror
+{
+    /// <summary>
+    /// Functions to Compress Quaternions and Floats
+    /// </summary>
+    /// <remarks>
+    /// Uncompressed Quaternion = 32 * 4 = 128 bits => send 16 bytes
+    ///
+    /// <para>
+    ///     Quaternion is always normalized so we drop largest value and re-calculate it.
+    ///     We can encode which one is the largest using 2 bits
+    ///     <code>
+    ///     x^2 + y^2 + z^2 + w^2 = 1
+    ///     </code>
+    /// </para>
+    ///
+    /// <para>
+    ///     2nd largest value has max size of 1/sqrt(2)
+    ///     We can encode the smallest three components in [-1/sqrt(2),+1/sqrt(2)] instead of [-1,+1]
+    ///     <code>
+    ///     c^2 + c^2 + 0 + 0 = 1
+    ///     </code>
+    /// </para>
+    /// 
+    /// <para>
+    ///     Sign of largest value doesnt matter
+    ///     <code>
+    ///     Q * vec3 == (-Q) * vec3
+    ///     </code>
+    /// </para>
+    /// 
+    /// <list type="bullet">
+    /// <listheader><description>
+    ///     RotationPrecision <br/>
+    ///     <code>
+    ///     2/sqrt(2) / (2^bitCount - 1)
+    ///     </code>
+    /// </description></listheader>
+    /// 
+    /// <item><description>
+    ///     rotation precision +-0.00138 in range [-1,+1]
+    ///     <code>
+    ///     10 bits per value
+    ///     2 + 10 * 3 = 32 bits => send 4 bytes
+    ///     </code>
+    /// </description></item>
+    /// </list>
+    ///
+    /// <para>
+    /// Links for more info:
+    /// <br/><see href="https://youtu.be/Z9X4lysFr64">GDC Talk</see>
+    /// <br/><see href="https://gafferongames.com/post/snapshot_compression/">Post on Snapshot Compression</see>
+    /// </para>
+    /// </remarks>
+    public static class Compression
+    {
+        const float QuaternionMinValue = -1f / 1.414214f; // 1/ sqrt(2)
+        const float QuaternionMaxValue = 1f / 1.414214f;
+
+        const int QuaternionBitLength = 10;
+        // same as Mathf.Pow(2, targetBitLength) - 1
+        const uint QuaternionUintRange = (1 << QuaternionBitLength) - 1;
+
+        /// <summary>
+        /// Used to Compress Quaternion into 4 bytes
+        /// </summary>
+        public static uint CompressQuaternion(Quaternion value)
+        {
+            value = value.normalized;
+
+            int largestIndex = FindLargestIndex(value);
+            Vector3 small = GetSmallerDimensions(largestIndex, value);
+            // largest needs to be positive to be calculated by reader 
+            // if largest is negative flip sign of others because Q = -Q
+            if (value[largestIndex] < 0)
```

Comment:
is there a test to confirm that this works?

Reply:
The `Q=-Q` part is written in full is `Q * vec3 == (-Q) * vec3` (which is in the remarks) It should just be the mathematical property of the  Quaternion. But I will add tests multiplying the before/after with a vector to make sure , and to check stuff if we ever change this.

Reply:
please do :)